discord.go

```diff
@@ -96,35 +98,202 @@ func (dm *DiscordManager) SaveServerOptions(oldServerOpts *GlobalServerConfig, n
+// The role will be added as a mention on all Discord notifications
+func (dm *DiscordManager) CommandNotify(s *discordgo.Session, m *discordgo.MessageCreate) (string, error) {
+	serverOpts, err := dm.store.LoadServerOptions()
+
+	if err != nil {
+		return "A server error occurred, try again later", err
 	}
 
-	if m.Content == "!schedule" {
-		start := time.Now()
-		end := start.AddDate(0, 0, 7)
+	if serverOpts.DiscordRoleID == "" || serverOpts.DiscordRoleCommand == "" {
+		return "", nil
+	}
 
-		calendar, err := dm.scheduledRacesManager.buildCalendar(start, end)
+	// get the member
+	member, err := s.State.Member(m.GuildID, m.Author.ID)
+
+	if err != nil {
+		// if it's not in the state, punt to asking the server
+		if err == discordgo.ErrStateNotFound {
+			member, err = s.GuildMember(m.GuildID, m.Author.ID)
+		}
 
 		if err != nil {
-			return
+			return "You don't seem to exist, so I can't assign you that role.  Try again later.", err
 		}
+	}
 
-		var msg = ""
+	// get the role name from ID, for use in user feedback
+	roleName := "notification"
+	roles, err := s.GuildRoles(m.GuildID)
 
-		for _, event := range calendar {
-			msg += event.Start.Format("Mon, 02 Jan 2006 15:04:05 MST") + "\n"
-			msg += event.Title + "\n"
-			msg += event.Description + "\n\n"
+	if err != nil {
+		// meh, just log it and carry on
+		logrus.WithError(err).Infof("failed to get Discord roles (make sure you have set the bot permissions to see roles)")
+	} else {
+		for _, role := range roles {
+			if strings.TrimSpace(role.ID) == strings.TrimSpace(serverOpts.DiscordRoleID) {
+				roleName = role.Name
+				break
+			}
 		}
+	}
 
-		_, err = s.ChannelMessageSend(m.ChannelID, msg)
+	for _, roleID := range member.Roles {
+		if roleID == serverOpts.DiscordRoleID {
+			// they have the role, so remove it
+			err = s.GuildMemberRoleRemove(m.GuildID, m.Author.ID, serverOpts.DiscordRoleID)
 
-		if err != nil {
-			logrus.Errorf("couldn't open discord session, err: %s", err)
+			if err != nil {
+				// meh, log the error here, and just return some feedback to the user
+				logrus.WithError(err).Infof("failed to remove Discord role (make sure you have set the bot permissions to manage roles)")
+				return fmt.Sprintf("You already have the %s role, and an error occurred trying to remove it.  Try again later.", roleName), nil
+			}
+
+			return fmt.Sprintf("You already had the %s role, it has now been removed.  Type the command again to add it back.", roleName), nil
 		}
 	}
+
+	// they didn't have the role, so add it
+	err = s.GuildMemberRoleAdd(m.GuildID, m.Author.ID, serverOpts.DiscordRoleID)
+
+	if err != nil {
+		// meh, log the error here and return feedback to the user
+		logrus.WithError(err).Infof("failed to set Discord role (make sure you have set the bot permissions to manage roles)")
+		return fmt.Sprintf("A server error occurred trying to assign you the %s role, try again later", roleName), nil
+	}
+
+	// w00t!
```

Comment:
:tada: :grin:

Reply:
There was much celebrating when I finally got to that line in testing.  :)